### ISSUE TITLE
Improve error messages

### ISSUE THREAD

**Issue by vladaionescu** [earthly/earthly#3200]
Some errors are difficult to parse by a human. Here are some examples...

Note that GH's representation doesn't do justice because there's no wrapping. In a terminal, the really long lines take up a LOT of real-estate space that looks very intimidating...

### Example 1

```
ERROR linkchecker/Earthfile line 46:6
  ./linkchecker+test *failed* |       The command
  ./linkchecker+test *failed* |           WITH DOCKER RUN --privileged docker-compose up --abort-on-container-exit
  ./linkchecker+test *failed* |       did not complete successfully. Exit code 1
Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev/.
Error: build target: build main: failed to solve: linkchecker/Earthfile line 46:6 with docker run: unlazy force execution: process "/bin/sh -c EARTHLY_DOCKERD_DATA_ROOT=\"/var/earthly/dind/16a66f7cc7947118c7b820248d045d48bc25b32a7ef4c8d7d4a35b036f1f748c\" EARTHLY_DOCKER_LOAD_FILES=\"\" EARTHLY_IMAGES_WITH_DIGESTS=\"package-manager:docs-linkchecker@sha256:a24601ea37a1705c5099f8852612ce8301c6c3af125e965b102f51c90e6749a6 package-manager:linkchecker@sha256:33143bb82a1b0b39ada222c12418480f1f1d354f1b5c1ccafbb0ffafabd14f84\" EARTHLY_START_COMPOSE=\"true\" EARTHLY_COMPOSE_FILES=\"compose.yml\" EARTHLY_COMPOSE_SERVICES=\"docs\" DIND_COMMIT=42b1175eda071c0e9121e1d64345928384a93df1 DOCKER_TLS_CERTDIR=/certs DOCKER_VERSION=20.10.14 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin build_number=development type=dev /var/earthly/dockerd-wrapper.sh execute /usr/bin/earth_debugger /bin/sh -c 'docker-compose up --abort-on-container-exit'" did not complete successfully: exit code: 1
in		github.com/rstudio/package-manager/docs/main/linkchecker:HEAD+test --build_number=development --type=dev

Error: Process completed with exit code 1.
```

##### Suggested improvement

This part is great:

```
ERROR linkchecker/Earthfile line 46:6
  ./linkchecker+test *failed* |       The command
  ./linkchecker+test *failed* |           WITH DOCKER RUN --privileged docker-compose up --abort-on-container-exit
  ./linkchecker+test *failed* |       did not complete successfully. Exit code 1
```

The line should be removed, or minimized drastically: ```Error: build target ...```

### Example 2

```
Error: build target: build main: failed to solve: Earthfile line 189:4 with docker run: unlazy force execution: error calling LocalhostExec: exit code: 100
in		github.com/rstudio/package-manager/docs/main:ah/install-migration+dev --port= --type=on_prem
```

This seems like an Earthly bug?

### Example 3

```
+insiders(https://github.com/rstudio/mkdocs-material-insiders)(https://github.com/rstudio/mkdocs-material-insiders) | Initialized empty Git repository in /tmp/earthly/buildkit/runc-overlayfs/snapshots/snapshots/88/fs/
+insiders(https://github.com/rstudio/mkdocs-material-insiders)(https://github.com/rstudio/mkdocs-material-insiders) | fatal: could not read Username for 'https://github.com/': terminal prompts disabled
+insiders(https://github.com/rstudio/mkdocs-material-insiders)(https://github.com/rstudio/mkdocs-material-insiders) | WARN: The command 'GIT CLONE (--branch 9.1.5-insiders-4.32.4) https://github.com/rstudio/mkdocs-material-insiders' failed: failed to fetch remote https://github.com/rstudio/mkdocs-material-insiders: exit status 128
../../scripts/earthfiles+node | WARN: Canceled
Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev/.
Error: build target: build main: failed to solve: async earthfile2llb for +image: Earthfile line 217:2 copy artifact: apply build +build: earthfile2llb for +build: Earthfile line 209:2 apply FROM +pre-build: apply build +pre-build: earthfile2llb for +pre-build: Earthfile line 159:2 apply FROM +deps: apply build +deps: earthfile2llb for +deps: Earthfile line 45:2 copy artifact: apply build +insiders: earthfile2llb for +insiders: Earthfile line 147:2 apply FOR ... IN: non constant build arg read request: failed to load cache key: failed to fetch remote https://github.com/rstudio/mkdocs-material-insiders: exit status 128
in		github.com/rstudio/package-manager/docs/main:sj/merge-docs+insiders
in		github.com/rstudio/package-manager/docs/main:sj/merge-docs+deps
in		github.com/rstudio/package-manager/docs/main:sj/merge-docs+pre-build
in		github.com/rstudio/package-manager/docs/main:sj/merge-docs+build --is_production=true --type=on_prem
in		github.com/rstudio/package-manager/docs/main:sj/merge-docs+image --is_production=true
```

##### Suggested improvements

* The prefix of the log lines showing the target is really long. It should handle GIT CLONE contexts a little better
* Presumably the real issue here is something to do with git credentials? This should be more obvious.

### Example 4

```
../utilities/aws+dev.login *failed* | An error occurred (InvalidGrantException) when calling the CreateToken operation: 
../utilities/aws+dev.login *failed* | ERROR ../utilities/aws/Earthfile line 45:4
../utilities/aws+dev.login *failed* |       The command
../utilities/aws+dev.login *failed* |           RUN --no-cache aws sso login
../utilities/aws+dev.login *failed* |       did not complete successfully. Exit code 254
Share your logs with an Earthly account (experimental)! Register for one at https://ci.earthly.dev/.
Error: build target: build main: failed to solve: async force execution for ../utilities/aws+dev.login: unlazy force execution: process "/bin/sh -c EARTHLY_CI=false PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /usr/bin/earth_debugger /bin/sh -c 'aws sso login'" did not complete successfully: exit code: 254
~/team-package-manager
```

This is similar to example 1.

### Example 5

```
      ./grid/cli+run | time="2023-02-08T18:44:39Z" level=error msg="failed to read from ptmx: read /dev/ptmx: input/output error"
```

I think that this line is actually completely unrelated to the actual reason of build failure, but is causing the user to think that it might be.

**Comment by futile** [comment]
Stumbled on the following error message that I had to figure out:

```terminal
❯ earthly build --artifact=lambda +lambda-debug
Incorrect Usage: invalid boolean value "lambda" for -artifact: parse error

NAME:
   earthly build - Build an Earthly target

[...]
   --artifact, -a                  Output specified artifact; a wildcard (*) can be used to output all artifacts (default: false)
[...]

Error: invalid boolean value "lambda" for -artifact: parse error
```

Figured out that I had to do earthly build --artifact +lambda-debug/lambda, but wanted to report the confusing error-message :slightly_smiling_face:

The nicest would maybe be `earthly +lambda-debug/target` just Doing The Right Thing™? But just a suggestion I thought of :)

**Comment by vladaionescu** [comment]
Another very common example:

```
Error: build target: build main: failed to solve: async force execution for +my-target: unlazy force execution: failed to compute cache key: ...
```

The actual error is at the end of this really long string (e.g. "file not found"), but most people don't read that far and are intimidated by what "unlazy force execution" might mean.

**Comment by shepherdjerred** [comment]
Another case:

```
earthly +insiders
 Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)

 Build 🔧
————————————————————————————————————————————————————————————————————————————————

        ubuntu:jammy | --> Load metadata ubuntu:jammy linux/arm64
+insiders(https://github.com/squidfunk/mkdocs-material-insiders)(https://github.com/squidfunk/mkdocs-material-insiders) | --> GIT CLONE (--branch 9.2.7-insiders-4.40.3) https://github.com/squidfunk/mkdocs-material-insiders
           +insiders | --> FROM ubuntu:jammy
           +insiders | [----------] 100% resolve docker.io/library/ubuntu:jammy@sha256:aabed3296a3d45cede1dc866a24476c4d7e093aa806263c27ddaadbdce3c1054
           +insiders | *cached* --> WORKDIR /workspace
+insiders(https://github.com/squidfunk/mkdocs-material-insiders)(https://github.com/squidfunk/mkdocs-material-insiders) | fatal: could not read Username for 'https://github.com': terminal prompts disabled
+insiders(https://github.com/squidfunk/mkdocs-material-insiders)(https://github.com/squidfunk/mkdocs-material-insiders) | WARN: The command 'GIT CLONE (--branch 9.2.7-insiders-4.40.3) https://github.com/squidfunk/mkdocs-material-insiders' failed: failed to fetch remote https://github.com/squidfunk/mkdocs-material-insiders: EARTHLY_GIT_STDERR: Z2l0IC1jIHByb3RvY29sLmZpbGUuYWxsb3c9dXNlciAtLWdpdC1kaXIgL3RtcC9lYXJ0aGx5L2J1aWxka2l0L3J1bmMtb3ZlcmxheWZzL3NuYXBzaG90cy9zbmFwc2hvdHMvMTE2My9mcyBscy1yZW1vdGUgb3JpZ2luIDkuMi43LWluc2lkZXJzLTQuNDAuMwpmYXRhbDogY291bGQgbm90IHJlYWQgVXNlcm5hbWUgZm9yICdodHRwczovL2dpdGh1Yi5jb20nOiB0ZXJtaW5hbCBwcm9tcHRzIGRpc2FibGVkCg==: exit status 128
Shareable link: https://ci.earthly.dev/logs?logId=27621cdd-6f65-4006-9e20-dbe6db10c47b
Error: build target: build main: failed to solve: async force execution for +insiders: unlazy force execution: failed to load cache key: failed to fetch remote https://github.com/squidfunk/mkdocs-material-insiders:  exit status 128

git -c protocol.file.allow=user --git-dir /tmp/earthly/buildkit/runc-overlayfs/snapshots/snapshots/1163/fs ls-remote origin 9.2.7-insiders-4.40.3
fatal: could not read Username for 'https://github.com': terminal prompts disabled
Check your git auth settings.
Did you ssh-add today? Need to configure ~/.earthly/config.yml?
For more information see https://docs.earthly.dev/guides/auth
```

The above error appears when Git credentials are incorrect. The error message is very informative, but _way_ too long.

**Comment by alexcb** [comment]
we should make a link to qemu easier for new users.

e.g. if `/usr/bin/earth_debugger: Invalid ELF image for this architecture` appears, we should include a link to https://docs.earthly.dev/docs/guides/multi-platform and mention they need to install qemu.

**Comment by idodod** [comment]
errors relating to invalid or missing target are showing as:
```
_unknown *failed* | no target reference provided. Try earthly +<target-name>
```

and 
```
_unknown *failed* | parse target name some-target: invalid target ref some-target
```

can be shown like a normal cli flag error without the prefix, i.e.:
```
no target reference provided. Try earthly +<target-name>
```

and 
```
parse target name some-target: invalid target ref some-target
```

**Comment by idodod** [comment]
Error launching a satellite
```
earthly sat launch my-satellite
Launching Satellite "my-satellite" with auto-updates set to run at 02:00 (BST)
This may take a few minutes...
Error: plan only allows 0 satellites in use at one time {reqID: ...}
            _unknown *failed* | failed to create satellite my-satellite: failed launching satellite: rpc error: code = PermissionDenied desc = rpc error: code = PermissionDenied desc = plan only allows 0 satellites in use at one time {reqID: ...}
```

**Comment by idodod** [comment]
> Another very common example:
> 
> ```
> Error: build target: build main: failed to solve: async force execution for +my-target: unlazy force execution: failed to compute cache key: ...
> ```
> 
> The actual error is at the end of this really long string (e.g. "file not found"), but most people don't read that far and are intimidated by what "unlazy force execution" might mean.

We already handle properly a `file not found` error, but I think other cases are a bug that we need to address before we can improve the error message, see #3150

**Comment by idodod** [comment]
This issue was resolved (see linked PRs) and are included up to latest release v0.7.21.
unaddressed issues now have separate issues (see links above as well)
